### DISCUSSION TITLE
Change background color of elements

### DISCUSSION THREAD

**Discussion by psinger** [h2oai/wave#1000] [Q&A]
Is it possible to change the background color of ui elements, such as `form_card`?

**Comment by mturoci** [comment]
Hi @psinger, this is not possible currently. Working on custom themes though, see https://github.com/h2oai/wave/pull/604. Note that it won't allow you to control color of individual cards.

**Reply by psinger** [reply]
Would this be very hard to implement? It sounds like a handy addition to being able to customize style aspects of individual items.

**Comment by Far0n** [comment]
@psinger If you want more control over the style of your app, you could also go down the css(/js)-path. Once the boilerplate is in place it works quite nicely, for example:
![grafik](https://user-images.githubusercontent.com/6949295/133925450-5b218005-2c40-4546-b31f-2358b31cd813.png)

**Reply by Far0n** [reply]
speaking of which: @mturoci would it be possible to add some functionalitiy to card creation that allow to specify html-attributes? Something like: 

```html
ui.form_card(... , tag="foobar")
=> <div ... tag="foobar"></div>
```
or even more general
```html
ui.form_card(... , html_attr={"tag": "foobar", "my_attr": "blub"})
=> <div ... tag="foobar" my_attr="blub"></div>
```

**Reply by mturoci** [reply]
@Far0n It is possible, but the goal of Wave is to allow people to build web apps without HTML / CSS / JS knowledge. Having such level of control kinda defeats the purpose of building web apps with Wave. It is better to build the web app in React if you happen to know HTML / CSS / JS anyway since it gives you 100% control. Even though it is possible to use HTML / CSS / JS in Wave, consider it just an escape hatch.

**Reply by Far0n** [reply]
@mturoci I only see benefits tbh, because it just adds additional functionality. I know, that it's not the intended way to use wave, but CSS/JS enable lots of possibilities for implementing advanced features while profiting from all the benefits wave offers. Features like above can still be advertised as "escapes hatches" so that the "advanced users" can make use of them while the general message of CSS/JS free app creation stays the same.

**Reply by mturoci** [reply]
I see your point @Far0n, but I wouldn't see this as a feature for advanced users still. A few downsides I see if implemented:

* the main (and probably only) reason we even support the HTML / CSS / JS is to allow people integrate 3rd party components like some fancy visualizations or tables, not to create own Wave-like components.
* maintainability - no matter how much advanced one is, having some parts of the app in python and some in native web technologies will inevitably become a mess sooner or later
* Hybrid solutions like the one mentioned above usually fail due to trying to solve everything, ending up a giant mess instead.
* Although we support tinkering with JS / HTML / CSS as an escape hatch only, this does not prevent people from asking more HTML control. As a result, the apps made by mixing native wave and custom HTML doesn't look nice and synchronized giving the impression Wave is not able to produce slick interfaces, which is basically one of our main goals.
* If we keep adding support for HTML attributes, the python code will soon look like some weird Frankenstein HTML which again makes no sense. It's better to pick one tech stack and stick with it.

Better way would be to file an issue when you feel like the component you need is missing and more people could also benefit from it.

cc @lo5 for more clarification - maybe we should write a blog post explaining this since it is a common question and I answered it a few times already at different communication mediums.

**Reply by Far0n** [reply]
gotcha @mturoci

**Reply by psinger** [reply]
@Far0n that example looks cool, do you have code for that? Curious how it works in detail. Are you also able to trigger actions from that?

**Reply by Far0n** [reply]
@psinger You can find it here: https://github.com/h2oai/corner-faron/tree/main/wave-recipe-builder/src
esp: https://github.com/h2oai/corner-faron/blob/21e91dfd95b3ecd0d06fa5c214f9a680765e1860/wave-recipe-builder/src/cards.py#L27

The idea is to use a small js-injector to get css into the page: https://github.com/h2oai/corner-faron/blob/21e91dfd95b3ecd0d06fa5c214f9a680765e1860/wave-recipe-builder/src/js/inject_css.js
However, that might be obsolete since wave 0.18 (https://github.com/h2oai/wave/blob/master/py/examples/meta_stylesheet.py)

The upside is, that the same approach works for all 3 - html, js and css and hence allows for some nice extra flexibility.

**Reply by Far0n** [reply]
> Are you also able to trigger actions from that?

Yes, for example each card acts like a button. Or what do you mean by "action triggering"?

https://user-images.githubusercontent.com/6949295/134041999-6acd5dd9-2eb1-43b3-a9e4-779c74286a1a.mp4

related: https://github.com/h2oai/wave/issues/945#issuecomment-895065239

**Reply by psinger** [reply]
Thanks @Far0n - cant access your github repo though :) but that looks like what I meant
